src/main/java/appeng/parts/p2p/PartP2PLiquids.java

```diff
@@ -254,10 +262,19 @@ public FluidStack drain( final ForgeDirection from, final int maxDrain, final bo
 		return null;
 	}
 
+	private boolean inputAcceptsFluid(Fluid input)
+	{
+		PartP2PLiquids l = getInput();
+		if (l == null || l == this)
+			return false;
+		final IFluidHandler handler = l.getTarget();
+		return handler != null && handler.canFill(l.getSide().getOpposite(), input);
+	}
+
 	@Override
 	public boolean canFill( final ForgeDirection from, final Fluid fluid )
 	{
-		return !this.isOutput() && from == this.getSide() && !this.getOutputs( fluid ).isEmpty();
+		return from == this.getSide() && (!this.getOutputs( fluid ).isEmpty() || inputAcceptsFluid(fluid));
```

Comment:
What do you think about splitting this into cars to clarify what you are testing?